[RFC] vim-patch:8.0.0{671,714,720,722,948,1020,1048,1507}

janlazo opened the pull request:
**vim-patch:8.0.0671: hang when typing CTRL-C in confirm() in timer**

Problem:    When a function invoked from a timer calls confirm() and the user
            types CTRL-C then Vim hangs.
Solution:   Reset typebuf_was_filled. (Ozaki Kiichi, closes vim/vim#1791)
https://github.com/vim/vim/commit/4eb6531b03445b4d492bc52fea0b6dcd886583af

 **vim-patch:8.0.0714: when a timer causes a command line redraw " goes missing**

Problem:    When a timer causes a command line redraw the " that is displayed
            for CTRL-R goes missing.
Solution:   Remember an extra character to display.
vim/vim@a92522f

**vim-patch:8.0.0720: unfinished mapping not displayed when running timer**

Problem:    Unfinished mapping not displayed when running timer.
Solution:   Also use the extra_char while waiting for a mapping and digraph.
            (closes vim/vim#1844)
vim/vim@6a77d26

**vim-patch:8.0.0722: screen is messed by timer up at inputlist() prompt**

Problem:    Screen is messed by timer up at inputlist() prompt.
Solution:   Set state to ASKMORE. (closes vim/vim#1843)
vim/vim@c904107

 **vim-patch:8.0.0948: crash if timer closes window while dragging status line**

Problem:    Crash if timer closes window while dragging status line.
Solution:   Check if the window still exists. (Yasuhiro Matsumoto, closes
            vim/vim#1979)
vim/vim@989a70c

**vim-patch:8.0.1020: when a timer calls getchar(1) input is overwritten**

Problem:    When a timer calls getchar(1) input is overwritten.
Solution:   Increment tb_change_cnt in inchar(). (closes vim/vim#1940)
vim/vim@0f0f230

**vim-patch:8.0.1048: no test for what 8.0.1020 fixes**

Problem:    No test for what 8.0.1020 fixes.
Solution:   Add test_feedinput().  Add a test. (Ozaki Kiichi, closes vim/vim#2046)
vim/vim@5e80de3

**vim-patch:8.0.1507: timer test is a bit flaky**

Problem:    Timer test is a bit flaky.
Solution:   Add it to the list of flaky tests.
vim/vim@bfbea56

I assume patches 671, 714, 720, 722 are OK (but can be improved) so I added more patches. I omitted `test_feedinput` from patch 1048.

I didn't include patches 1424 and 1456 because `Test_paused` already has a longer duration (100 ms).

janlazo committed:
vim-patch:8.0.0671: hang when typing CTRL-C in confirm() in timer

Problem:    When a function invoked from a timer calls confirm() and the user
            types CTRL-C then Vim hangs.
Solution:   Reset typebuf_was_filled. (Ozaki Kiichi, closes vim/vim#1791)
https://github.com/vim/vim/commit/4eb6531b03445b4d492bc52fea0b6dcd886583af

janlazo committed:
vim-patch:8.0.0714: when a timer causes a command line redraw " goes missing

Problem:    When a timer causes a command line redraw the " that is displayed
            for CTRL-R goes missing.
Solution:   Remember an extra character to display.
https://github.com/vim/vim/commit/a92522fbf3a49d06e08caf010f7d7b0f58d2e131

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -194,6 +194,9 @@ static int cmd_showtail;                /* Only show path tail in lists ? */
 
 static int new_cmdpos;          /* position set by set_cmdline_pos() */
 
+static int extra_char = NUL;    // extra character to display when redrawing
```
Should this be a member of `command_line_state` struct?

mhinz replied:
I think we should keep consistent with Vim here to avoid future clashes (even if minimal).

bfredl replied:
There already is `ccline.special_char` (and `special_shift`), though it is only set for external cmdline currently. It could be changed to always set and used for this.

bfredl replied:
Though it could be renamed to `extra_char` for consistency with vim, it must be part of the struct as nvim supports redrawing of nested cmdline unlike vim.

janlazo replied:
Will both new static variables be moved to `command_line_state` or do I simply rename `special_char` and `special_shift` and use those instead? I can add more timer-related patches in the meantime while there is no resolution on this.

bfredl replied:
not `command_line_state` but `struct cmdline_info`, whose static instance `ccline` is the current commandline. In there `special_char` can be renamed `extra_char` for analogy with vim.

janlazo committed:
vim-patch:8.0.0720: unfinished mapping not displayed when running timer

Problem:    Unfinished mapping not displayed when running timer.
Solution:   Also use the extra_char while waiting for a mapping and digraph.
            (closes vim/vim#1844)
https://github.com/vim/vim/commit/6a77d2667e982655f6adacee774ee7aa2581bd8a

janlazo committed:
vim-patch:8.0.0722: screen is messed by timer up at inputlist() prompt

Problem:    Screen is messed by timer up at inputlist() prompt.
Solution:   Set state to ASKMORE. (closes vim/vim#1843)
https://github.com/vim/vim/commit/c9041079a199d753e73d3b242f21cc8db620179a

mhinz commented on src/nvim/ex_getln.c:
```diff
@@ -3470,6 +3479,9 @@ void redrawcmd(void)
 
   set_cmdspos_cursor();
+  if (extra_char != NUL) {
+    putcmdline(extra_char, true);
```
We could change the signature of `putcmdline()` to `void putcmdline(int c, bool shift)` in `ex_getln.c` as well.

mhinz reviewed (approved):
Apart from the nitpick of changing the signature to `void putcmdline(int c, bool shift)`, the patches LGTM.

janlazo committed:
vim-patch:8.0.0948: crash if timer closes window while dragging status line

Problem:    Crash if timer closes window while dragging status line.
Solution:   Check if the window still exists. (Yasuhiro Matsumoto, closes
            vim/vim#1979)
https://github.com/vim/vim/commit/989a70c590c2bd109eb362d3a0e48cb1427ae13d

janlazo committed:
vim-patch:8.0.1020: when a timer calls getchar(1) input is overwritten

Problem:    When a timer calls getchar(1) input is overwritten.
Solution:   Increment tb_change_cnt in inchar(). (closes vim/vim#1940)
https://github.com/vim/vim/commit/0f0f230012f5a9beb6876158a17b432534836c6f

janlazo committed:
vim-patch:8.0.1507: timer test is a bit flaky

Problem:    Timer test is a bit flaky.
Solution:   Add it to the list of flaky tests.
https://github.com/vim/vim/commit/bfbea567d89fdaa08ed987fd80daa53a6ce399d1

janlazo committed:
vim-patch:8.0.1048: no test for what 8.0.1020 fixes

Problem:    No test for what 8.0.1020 fixes.
Solution:   Add test_feedinput().  Add a test. (Ozaki Kiichi, closes vim/vim#2046)
https://github.com/vim/vim/commit/5e80de3f3e572805fe734b66bc42c13303ad9bdb

janlazo commented:
I added patches that had `timer` in the commit message. Since `putcmdline` is untouched by those patches, I'll change the signature after trying @bfredl suggestion with the static `ccline`.

janlazo commented:
Why did the test fail?

```
[  ERROR   ] .../build/neovim/neovim/test/functional/ui/cmdline_spec.lua @ 276: external cmdline works with special chars and nested cmdline
test/functional/ui/screen.lua:305: .../build/neovim/neovim/test/functional/ui/cmdline_spec.lua:360: Expected objects to be the same.
Passed in:
(table) {
 *[1] = {
    [content] = {
      [1] = {
        [1] = { }
        [2] = 'xx' } }
    [firstc] = ':'
    [indent] = 0
    [pos] = 2
    [prompt] = '' }
  [2] = {
    [content] = {
      [1] = {
        [1] = { }
        [2] = '1' }
      [2] = {
        [1] = { }
        [2] = '+' }
      [3] = {
        [1] = { }
        [2] = '2' } }
    [firstc] = '='
    [indent] = 0
    [pos] = 3
    [prompt] = '' } }
Expected:
(table) {
 *[1] = {
    [content] = {
      [1] = {
        [1] = { }
        [2] = 'xx' } }
    [firstc] = ':'
    [indent] = 0
    [pos] = 2
    [prompt] = ''
   *[special] = {
      [1] = '"'
      [2] = true } }
  [2] = {
    [content] = {
      [1] = {
        [1] = { }
        [2] = '1' }
      [2] = {
        [1] = { }
        [2] = '+' }
      [3] = {
        [1] = { }
        [2] = '2' } }
    [firstc] = '='
    [indent] = 0
    [pos] = 3
    [prompt] = '' } }
stack traceback:
	test/functional/ui/screen.lua:305: in function 'wait'
	test/functional/ui/screen.lua:213: in function 'expect'
	.../build/neovim/neovim/test/functional/ui/cmdline_spec.lua:353: in function <.../build/neovim/neovim/test/functional/ui/cmdline_spec.lua:276>
```

janlazo committed:
ex_getln: update putcmdline(int, bool) signature

janlazo commented:
@bfredl Can you make the change to use `ccline` for `extra_char` and `extra_char_shift`? When I did it, a test failed,

janlazo commented:
I can rename `special_char` and `special_shift` but I don't know how to make the test pass so I'll leave this pending for some time. Maybe after https://github.com/neovim/neovim/pull/8605.
